modules/Bio/EnsEMBL/Compara/RunnableDB/GeneTrees/OrthoTree.pm

```diff
@@ -566,5 +602,115 @@ sub check_homology_consistency {
+
+    my $c = 1;
+    my %seen_homologies;
+    while ( my $line = <$fh> ) {
+        my $row = map_row_to_header($line, \@header_cols);
+        my ( $st_id, $hom_st_id ) = ($row->{stable_id}, $row->{homology_stable_id});
+
+        # Each homology must be linked to exactly 2 members
+        unless (defined $st_id && defined $hom_st_id) {
+            die sprintf(
+                "Homology should have 2 members: %s & %s (line %d)\n",
+                ( $st_id, $hom_st_id, $c )
+            );
+        }
+
+        # A pair of gene can only appear in 1 homology at most
+        $seen_homologies{"$st_id - $hom_st_id"}++;
+        $seen_homologies{"$hom_st_id - $st_id"}++;
+        if ($seen_homologies{"$st_id - $hom_st_id"} > 1 ||
+            $seen_homologies{"$hom_st_id - $st_id"} > 1 ) {
+            die sprintf(
+                "Pair (%s - %s) seen more than once (line %d)",
+                ( $st_id, $hom_st_id, $c )
+            );
+        }
+
+        # Checks that all the relevant fields are non-NULL or non-zero
+        unless (
+            defined $row->{homology_type} &&
+            defined $row->{seq_member_id} &&
+            defined $row->{homology_seq_member_id} &&
+            defined $row->{cigar_line} &&
+            length($row->{cigar_line}) > 0 &&
+            defined $row->{homology_cigar_line} &&
+            length($row->{homology_cigar_line}) > 0 &&
+            defined $row->{perc_id} &&
+            defined $row->{homology_perc_id} &&
+            defined $row->{perc_pos} &&
+            defined $row->{homology_perc_pos}
+        ) {
+            die sprintf(
+                "Homology is missing key values ($st_id - $hom_st_id) (line %d)",
+                ( $st_id, $hom_st_id, $c )
+            );
+        }
+
+        # Checks that the seq_member_id only links to canonical peptides
+        my $gene_member = $gene_member_adaptor->fetch_by_dbID($row->{gene_member_id});
+        my $hom_gene_member = $gene_member_adaptor->fetch_by_dbID($row->{homology_gene_member_id});
+        if ( $gene_member->canonical_member_id != $row->{seq_member_id} ) {
+            die sprintf(
+                "%s is not a canonical member (line %d)",
+                ($st_id, $c)
+            );
+        } elsif ( $hom_gene_member->canonical_member_id != $row->{homology_seq_member_id} ) {
+            die sprintf(
+                "%s is not a canonical member (line %d)",
+                ($hom_st_id, $c)
+            );
+        }
+
+        $c++;
+    }
+}
+
+sub _check_file_integrity {
+    my ( $self, $file ) = @_;
+
+    # check line counts
+    my $leaf_count = scalar @{ $self->param('gene_tree')->root->get_all_leaves };
+    my $exp_lines = (($leaf_count * ($leaf_count - 1)) / 2) + 1; # add 1 for header line
+    my @wc_output = split(/\s+/, $self->get_command_output("wc -l $file"));
+    my $got_line_count = $wc_output[0];
+    die "Expected $exp_lines lines, but got $got_line_count: $file" if $exp_lines != $got_line_count;
```

Comment:
That'd be useful in `Utils::FlatFile` and could be used in RunnableDB/GeneTrees/DumpHomologiesTSV.pm